File: app/src/main/java/io/github/studerus/pepper_android_realtime/tools/information/GetDateTimeTool.java
package io.github.studerus.pepper_android_realtime.tools.information;

import io.github.studerus.pepper_android_realtime.tools.Tool;
import io.github.studerus.pepper_android_realtime.tools.ToolContext;

import org.json.JSONObject;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

/**
 * Tool for getting current date and time with various formatting options.
 * Supports custom formatting, timezones, and locale-specific formatting.
 */
public class GetDateTimeTool implements Tool {

    @Override
    public String getName() {
        return "get_current_datetime";
    }

    @Override
    public JSONObject getDefinition() {
        try {
            JSONObject tool = new JSONObject();
            tool.put("type", "function");
            tool.put("name", getName());
            tool.put("description", "Get the current date and time. Supports optional format and timezone.");
            
            JSONObject params = new JSONObject();
            params.put("type", "object");
            
            JSONObject properties = new JSONObject();
            properties.put("format", new JSONObject()
                .put("type", "string")
                .put("description", "Formatting style. Default is iso."));
            properties.put("timezone", new JSONObject()
                .put("type", "string")
                .put("description", "IANA timezone, e.g. Europe/Zurich."));
            properties.put("pattern", new JSONObject()
                .put("type", "string")
                .put("description", "Java SimpleDateFormat pattern."));
            
            params.put("properties", properties);
            tool.put("parameters", params);
            
            return tool;
        } catch (Exception e) {
            throw new RuntimeException("Failed to create tool definition", e);
        }
    }

    @Override
    public String execute(JSONObject args, ToolContext context) throws Exception {
        String format = args.optString("format", "iso");
        String timezone = args.optString("timezone", "");
        String pattern = args.optString("pattern", "");

        TimeZone tz = timezone.isEmpty() ? TimeZone.getDefault() : TimeZone.getTimeZone(timezone);
        long epochMillis = System.currentTimeMillis();
        Date now = new Date(epochMillis);
        
        String formatted;
        if ("locale".equalsIgnoreCase(format)) {
            java.text.DateFormat df = java.text.DateFormat.getDateTimeInstance(
                    java.text.DateFormat.MEDIUM,
                    java.text.DateFormat.MEDIUM);
            df.setTimeZone(tz);
            formatted = df.format(now);
        } else if ("custom".equalsIgnoreCase(format) && !pattern.isEmpty()) {
            SimpleDateFormat sdf = new SimpleDateFormat(pattern, Locale.US);
            sdf.setTimeZone(tz);
            formatted = sdf.format(now);
        } else {
            formatted = formatAsRfc3339(now, tz);
        }
        
        JSONObject result = new JSONObject();
        result.put("datetime", formatted);
        result.put("epochMillis", epochMillis);
        result.put("timezone", tz.getID());
        
        return result.toString();
    }

    @Override
    public boolean requiresApiKey() {
        return false;
    }

    @Override
    public String getApiKeyType() {
        return null;
    }

    /**
     * Format date as RFC3339 string
     */
    private String formatAsRfc3339(Date date, TimeZone timeZone) {
        @SuppressWarnings("SpellCheckingInspection") // yyyy-MM-dd'T'HH:mm:ssZ is correct RFC3339 format
        SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.US);
        sdf.setTimeZone(timeZone);
        String raw = sdf.format(date);
        if (raw.length() > 5) {
            return raw.substring(0, raw.length() - 2) + ":" + raw.substring(raw.length() - 2);
        }
        return raw;
    }
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/tools/information/GetDateTimeTool.kt
package io.github.anonymous.pepper_realtime.tools.information

import io.github.anonymous.pepper_realtime.tools.Tool
import io.github.anonymous.pepper_realtime.tools.ToolContext
import org.json.JSONObject
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Tool for getting current date and time with various formatting options.
 * Supports custom formatting, timezones, and locale-specific formatting.
 */
class GetDateTimeTool : Tool {

    override fun getName(): String = "get_current_datetime"

    override fun getDefinition(): JSONObject {
        return JSONObject().apply {
            put("type", "function")
            put("name", getName())
            put("description", "Get the current date and time. Supports optional format and timezone.")

            put("parameters", JSONObject().apply {
                put("type", "object")
                put("properties", JSONObject().apply {
                    put("format", JSONObject()
                        .put("type", "string")
                        .put("description", "Formatting style. Default is iso."))
                    put("timezone", JSONObject()
                        .put("type", "string")
                        .put("description", "IANA timezone, e.g. Europe/Zurich."))
                    put("pattern", JSONObject()
                        .put("type", "string")
                        .put("description", "Java SimpleDateFormat pattern."))
                })
            })
        }
    }

    override fun execute(args: JSONObject, context: ToolContext): String {
        val format = args.optString("format", "iso")
        val timezone = args.optString("timezone", "")
        val pattern = args.optString("pattern", "")

        val tz = if (timezone.isEmpty()) TimeZone.getDefault() else TimeZone.getTimeZone(timezone)
        val epochMillis = System.currentTimeMillis()
        val now = Date(epochMillis)

        val formatted = when {
            format.equals("locale", ignoreCase = true) -> {
                val df = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.MEDIUM)
                df.timeZone = tz
                df.format(now)
            }
            format.equals("custom", ignoreCase = true) && pattern.isNotEmpty() -> {
                val sdf = SimpleDateFormat(pattern, Locale.US)
                sdf.timeZone = tz
                sdf.format(now)
            }
            else -> formatAsRfc3339(now, tz)
        }

        return JSONObject().apply {
            put("datetime", formatted)
            put("epochMillis", epochMillis)
            put("timezone", tz.id)
        }.toString()
    }


    /**
     * Format date as RFC3339 string
     */
    @Suppress("SpellCheckingInspection") // yyyy-MM-dd'T'HH:mm:ssZ is correct RFC3339 format
    private fun formatAsRfc3339(date: Date, timeZone: TimeZone): String {
        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ", Locale.US)
        sdf.timeZone = timeZone
        val raw = sdf.format(date)
        return if (raw.length > 5) {
            raw.substring(0, raw.length - 2) + ":" + raw.substring(raw.length - 2)
        } else {
            raw
        }
    }
}


